File: ios/Nexo/Models/ActivePlanItem.swift
//
//  ActivePlanItem.swift
//  Nexo
//

import Foundation

/// One item in a member's credit wallet — created when a `MembershipPlan`
/// is granted to them. Stored at
/// `users/{uid}/memberships/{gymId}/activePlans/{id}`.
struct ActivePlanItem: Codable, Hashable, Identifiable {
    var id: String
    var planName: String
    var type: PlanComponentType
    var resetPeriod: PlanResetPeriod
    /// nil means "all class types". Otherwise one of the gym's `workoutTypes`.
    var workoutType: String?
    /// Base credit allowance (e.g. 12 credits/month or 10 total credits).
    var creditCount: Int
    /// Remaining credits for fixed punch cards (`resetPeriod == .none`).
    var remainingCredits: Int
    /// How many credits have been consumed in the current monthly cycle (`resetPeriod == .monthly`).
    var cycleCreditsUsed: Int
    /// When the plan started / was granted, used as anchor for monthly cycle boundaries.
    var cycleAnchorDate: Date
    /// Month index when credits were last consumed (0 for month 1, 1 for month 2, etc.).
    var lastCycleIndex: Int
    /// Overall expiration date of the plan/contract.
    var expiresAt: Date

    init(
        id: String,
        planName: String,
        type: PlanComponentType = .unlimited,
        resetPeriod: PlanResetPeriod = .none,
        workoutType: String? = nil,
        creditCount: Int = 0,
        remainingCredits: Int = 0,
        cycleCreditsUsed: Int = 0,
        cycleAnchorDate: Date = Date(),
        lastCycleIndex: Int = 0,
        expiresAt: Date
    ) {
        self.id = id
        self.planName = planName
        self.type = type
        self.resetPeriod = resetPeriod
        self.workoutType = workoutType
        self.creditCount = creditCount
        self.remainingCredits = remainingCredits
        self.cycleCreditsUsed = cycleCreditsUsed
        self.cycleAnchorDate = cycleAnchorDate
        self.lastCycleIndex = lastCycleIndex
        self.expiresAt = expiresAt
    }

    var isExpired: Bool {
        expiresAt < Date()
    }

    /// Current cycle index relative to anchor date (0 for 1st month, 1 for 2nd month, etc.).
    func currentCycleIndex(relativeTo now: Date = Date()) -> Int {
        let calendar = Calendar.current
        let components = calendar.dateComponents([.month], from: cycleAnchorDate, to: now)
        return max(0, components.month ?? 0)
    }

    /// The start and end bounds of the current monthly reset cycle.
    func currentCycleBounds(relativeTo now: Date = Date()) -> (start: Date, end: Date) {
        let calendar = Calendar.current
        let index = currentCycleIndex(relativeTo: now)
        let start = calendar.date(byAdding: .month, value: index, to: cycleAnchorDate) ?? cycleAnchorDate
        let end = calendar.date(byAdding: .month, value: index + 1, to: cycleAnchorDate) ?? expiresAt
        return (start, min(end, expiresAt))
    }

    /// Available credits in the current period.
    func availableCredits(relativeTo now: Date = Date()) -> Int {
        guard !isExpired else { return 0 }
        switch type {
        case .unlimited:
            return 0
        case .credits:
            if resetPeriod == .monthly {
                let currentIndex = currentCycleIndex(relativeTo: now)
                let used = (currentIndex == lastCycleIndex) ? cycleCreditsUsed : 0
                return max(0, creditCount - used)
            } else {
                return max(0, remainingCredits)
            }
        }
    }

    /// Whether this item authorizes booking the given class. Premium classes can
    /// only be authorized by a component that explicitly names their `classType` —
    /// a generic "all classes" component (`workoutType == nil`) doesn't cover them.
    func matches(gymClass: GymClass) -> Bool {
        guard !isExpired else { return false }
        if gymClass.isPremium {
            return workoutType == gymClass.classType
        } else {
            return workoutType == nil || workoutType == gymClass.classType
        }
    }
}

File: android/app/src/main/java/com/nexo/app/domain/model/ActivePlanItem.kt
package com.nexo.app.domain.model

import java.time.Instant
import java.time.Period
import java.time.ZoneId

/**
 * One item in a member's credit wallet — created from a single
 * `PlanComponent` when a `MembershipPlan` is granted to them. Stored at
 * `users/{uid}/memberships/{gymId}/activePlans/{id}`. Mirrors iOS's
 * `ActivePlanItem`.
 */
data class ActivePlanItem(
    val id: String,
    val planName: String,
    val type: PlanComponentType,
    val resetPeriod: PlanResetPeriod = PlanResetPeriod.NONE,
    /** `null` means "all class types". Otherwise one of the gym's `workoutTypes`. */
    val workoutType: String? = null,
    /** Base credit allowance (e.g. 12 credits/month or 10 total credits). */
    val creditCount: Int = 0,
    /** Remaining credits for fixed punch cards ([resetPeriod] is [PlanResetPeriod.NONE]). */
    val remainingCredits: Int = 0,
    /** How many credits have been consumed in the current monthly cycle ([resetPeriod] is [PlanResetPeriod.MONTHLY]). */
    val cycleCreditsUsed: Int = 0,
    /** When the plan started / was granted, used as anchor for monthly cycle boundaries. */
    val cycleAnchorDateMillis: Long = System.currentTimeMillis(),
    /** Month index when credits were last consumed (0 for month 1, 1 for month 2, etc.). */
    val lastCycleIndex: Int = 0,
    val expiresAtMillis: Long
) {
    val isExpired: Boolean get() = expiresAtMillis < System.currentTimeMillis()

    /** Current cycle index relative to [cycleAnchorDateMillis] (0 for the 1st month, 1 for the 2nd, etc.). */
    fun currentCycleIndex(relativeToMillis: Long = System.currentTimeMillis(), zoneId: ZoneId = ZoneId.systemDefault()): Int {
        val anchor = Instant.ofEpochMilli(cycleAnchorDateMillis).atZone(zoneId).toLocalDate()
        val now = Instant.ofEpochMilli(relativeToMillis).atZone(zoneId).toLocalDate()
        val months = Period.between(anchor.withDayOfMonth(1), now.withDayOfMonth(1)).toTotalMonths().toInt()
        return maxOf(0, months)
    }

    /** The start and end bounds of the current monthly reset cycle. */
    fun currentCycleBounds(relativeToMillis: Long = System.currentTimeMillis(), zoneId: ZoneId = ZoneId.systemDefault()): Pair<Long, Long> {
        val index = currentCycleIndex(relativeToMillis, zoneId)
        val anchor = Instant.ofEpochMilli(cycleAnchorDateMillis).atZone(zoneId)
        val start = anchor.plusMonths(index.toLong()).toInstant().toEpochMilli()
        val end = anchor.plusMonths((index + 1).toLong()).toInstant().toEpochMilli()
        return start to minOf(end, expiresAtMillis)
    }

    /** Available credits in the current period. */
    fun availableCredits(relativeToMillis: Long = System.currentTimeMillis(), zoneId: ZoneId = ZoneId.systemDefault()): Int {
        if (isExpired) return 0
        return when (type) {
            PlanComponentType.UNLIMITED -> 0
            PlanComponentType.CREDITS -> {
                if (resetPeriod == PlanResetPeriod.MONTHLY) {
                    val currentIndex = currentCycleIndex(relativeToMillis, zoneId)
                    val used = if (currentIndex == lastCycleIndex) cycleCreditsUsed else 0
                    maxOf(0, creditCount - used)
                } else {
                    maxOf(0, remainingCredits)
                }
            }
        }
    }

    /**
     * Whether this item authorizes booking [gymClass] — mirrors iOS's
     * `ActivePlanItem.matches(gymClass:)`. A premium class can only be
     * authorized by an item that explicitly names its `classType`; a
     * generic "all classes" item (`workoutType == null`) doesn't cover it.
     */
    fun matches(gymClass: GymClass): Boolean {
        if (isExpired) return false
        return if (gymClass.isPremium) {
            workoutType == gymClass.classType
        } else {
            workoutType == null || workoutType == gymClass.classType
        }
    }
}
